### PR TITLE
[workloadmeta] New remote process collector

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR:
- Makes remote workload meta more generic and add unit tests.
- Adds the new remote process collector, . 

### Motivation
We need to collect process data in the core-agent for future use-cases. This data is currently stored in the process-agent.

### Possible Drawbacks / Trade-offs
Memory increase, uses a gRPC connection.

### Describe how to test/QA your changes
- Run unit tests. 
- Run the same test/QA as https://github.com/DataDog/datadog-agent/pull/14482 to make sure remote workloadmeta still works in the process agent. Processes must be tagged with container tags and live containers must work when remote workloadmeta is enabled.
- Deploy the agent with remote process collection enabled and make sure processes are collected with container tags following the description below.

1. Deploy the agent with helm and the following config:
```
datadog:
  env:
    - name: DD_PROCESS_CONFIG_LANGUAGE_DETECTION_ENABLED
      value: "true"
  processAgent:
    enabled: true
    processCollection: true
```

2. Edit the daemonset to insert `DD_WORKLOADMETA_REMOTE_PROCESS_COLLECTOR_ENABLED="true"` in the core agent (a helm chart update will arrive to make it easier)
3. Start a python process on a node
4. Make sure it is visible in the output of `agent workload-list -v`

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/workloadmeta/collectors/internal/remote/generic.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+// ResponseHandler handles a response sent by the remote gRPC endpoint
+type ResponseHandler func(response interface{}) ([]workloadmeta.CollectorEvent, error)
+
+type OnResync func(store workloadmeta.Store, events []workloadmeta.CollectorEvent)
+
+type GenericCollector struct {
+	NewClient       NewClient
+	ResponseHandler ResponseHandler
+	Port            int
+	OnResync        OnResync
```
Comment (clamoriniere) [inline pkg/workloadmeta/collectors/internal/remote/generic.go:57]: change the order

File: pkg/workloadmeta/collectors/internal/remote/generic.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+
+const (
+	noTimeout         = 0 * time.Minute
+	streamRecvTimeout = 10 * time.Minute
+)
+
+var errWorkloadmetaStreamNotStarted = errors.New("workloadmeta stream not started")
+
+// NewClient returns a remote grpc client
+type NewClient func(cc grpc.ClientConnInterface) RemoteGrpcClient
```
Comment (clamoriniere) [inline pkg/workloadmeta/collectors/internal/remote/generic.go:36]:
```suggestion
type NewClientFuncType func(cc grpc.ClientConnInterface) RemoteGrpcClient
```

**Review by just-chillin — APPROVED** [review]

**Comment by just-chillin** [comment]
Note that the test yaml is wrong. It should be:
```yaml
process_config:
  process_collection: {enabled: true}
  language_detection: {enabled: true}
```

**Comment by AliDatadog** [comment]
> Note that the test yaml is wrong. It should be:
> 
> ```yaml
> process_config:
>   process_collection: {enabled: true}
>   language_detection: {enabled: true}
> ```

I forgot to add that I was referring to the helm chart

File: pkg/util/proto/workloadmeta.go [hunk 3]
```diff
@@ -419,6 +421,53 @@ func WorkloadmetaFilterFromProtoFilter(protoFilter *pb.WorkloadmetaFilter) (*wor
...
+func toLanguage(proto *pb.Language) *languagemodels.Language {
+	if proto == nil {
+		return nil
+	}
+	return &languagemodels.Language{
+		Name: languagemodels.LanguageName(proto.GetName()),
+	}
+}
+
+// WorkloadmetaEventFromProcessEventSet converts the given ProcessEventSet into a workloadmeta.Event
```
Comment (vboulineau) [inline pkg/util/proto/workloadmeta.go:433]: Given that's not generic code, I'm not sure it should be in this package, perhaps more suited in the `process_collector`

**Review by vboulineau — APPROVED** [review]

**Review by just-chillin — APPROVED** [review]

**Review by urseberry — APPROVED** [review]

**Pull request merged**